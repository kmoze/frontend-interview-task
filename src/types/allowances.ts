export type Allowance = {
  id: number
  name: string
  renewal: 'week' | 'month' | 'year'
  currency: string
  amount: string
  spent: string
  active: boolean
  type: 'expense' | 'card'
}
