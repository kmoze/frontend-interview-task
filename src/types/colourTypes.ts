import { COLOUR_OPTIONS } from '@/constants/colourOptions'

export type ColourOption = (typeof COLOUR_OPTIONS)[keyof typeof COLOUR_OPTIONS]
