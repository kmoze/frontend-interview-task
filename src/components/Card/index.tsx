import React from 'react'
import { Allowance } from '@/types/allowances'
import { Inter } from 'next/font/google'

const inter = Inter({ subsets: ['latin'] })

type CardProps = {
  allowance: Allowance
}

const Card = ({ allowance }: CardProps) => {
  const spentPercentage =
    (parseFloat(allowance.spent) / parseFloat(allowance.amount)) * 100

  return (
    <div className="bg-white rounded-lg p-6 border border-[#DDDDDD] shadow-sm">
      <div className="mb-4 space-y-1.5">
        <h3
          className={`${inter.className} text-base font-medium text-[#232323]`}
        >
          {allowance.name}
        </h3>
        <p className={`${inter.className} text-sm text-[#797979] font-normal`}>
          {allowance.type === 'card' ? 'Spend Card' : 'Expense'}
        </p>
      </div>

      {allowance.active ? (
        <div className="space-y-[-0.25rem]">
          <div className="flex justify-between mb-2">
            <span
              className={`${inter.className} text-[#232323] text-sm font-medium`}
            >
              {Math.round(spentPercentage)}% utilised
            </span>
            <span className={`${inter.className} text-[#9A9A9A] text-sm`}>
              {allowance.currency}
              {allowance.amount} /{' '}
              {allowance.renewal.charAt(0).toUpperCase() +
                allowance.renewal.slice(1)}
            </span>
          </div>
          <div className="w-full h-1 bg-[#DDDDDD] rounded-full">
            <div
              className="h-full bg-[#64D98A] rounded-full"
              style={{ width: `${spentPercentage}%` }}
            />
          </div>
        </div>
      ) : (
        <button className={`${inter.className} text-[#569F6E] font-medium`}>
          Activate card
        </button>
      )}
    </div>
  )
}

export default Card
