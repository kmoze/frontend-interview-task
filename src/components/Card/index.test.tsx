import { render } from '@testing-library/react'
import Card from '@/components/Card'
import '@testing-library/jest-dom'
import { Allowance } from '@/types/allowances'

// Mock data for testing
const mockAllowanceActive: Allowance = {
  id: 1,
  name: 'Lunch',
  renewal: 'month',
  currency: '£',
  amount: '100',
  spent: '50',
  active: true,
  type: 'card',
}

const mockAllowanceInactive: Allowance = {
  id: 2,
  name: 'Team Day',
  renewal: 'year',
  currency: '£',
  amount: '200',
  spent: '0',
  active: false,
  type: 'card',
}

describe('Card Component', () => {
  it('renders correctly for active allowances', () => {
    const { getByText, getByTestId } = render(
      <Card allowance={mockAllowanceActive} />
    )

    expect(getByText('Lunch')).toBeInTheDocument()
    expect(getByText('Spend Card')).toBeInTheDocument()
    expect(getByText('50% utilised')).toBeInTheDocument()
    expect(getByText('£100 / month')).toBeInTheDocument()

    const progressBar = getByTestId('progress-bar')
    expect(progressBar).toBeInTheDocument()

    expect(progressBar).toHaveStyle('width: 50%')
  })

  it('renders correctly for inactive allowances', () => {
    const { getByText, queryByTestId } = render(
      <Card allowance={mockAllowanceInactive} />
    )

    expect(getByText('Team Day')).toBeInTheDocument()
    expect(getByText('Spend Card')).toBeInTheDocument()
    expect(getByText('Activate card')).toBeInTheDocument()
    expect(queryByTestId('progress-bar')).not.toBeInTheDocument()
  })

  it('renders utilisation bar width correctly', () => {
    const { getByTestId } = render(<Card allowance={mockAllowanceActive} />)

    const progressBar = getByTestId('progress-bar')
    expect(progressBar).toHaveStyle('width: 50%')
  })

  it('renders with a snapshot', () => {
    const { container } = render(<Card allowance={mockAllowanceActive} />)
    expect(container).toMatchSnapshot()
  })
})
