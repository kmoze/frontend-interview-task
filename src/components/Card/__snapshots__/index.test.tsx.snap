// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Component renders with a snapshot 1`] = `
<div>
  <div
    class="bg-white rounded-lg p-6 border border-[#DDDDDD] shadow-sm min-h-[185px] flex flex-col justify-between"
    data-testid="allowance-card"
  >
    <div
      class="space-y-1.5"
    >
      <h3
        class="className text-base font-medium text-[#232323]"
      >
        Lunch
      </h3>
      <p
        class="className text-sm text-[#797979] font-normal"
      >
        Spend Card
      </p>
    </div>
    <div
      class="space-y-[-0.25rem]"
    >
      <div
        class="flex justify-between mb-2"
      >
        <span
          class="className text-[#232323] text-sm font-medium"
        >
          50
          % utilised
        </span>
        <span
          class="className text-[#9A9A9A] text-sm"
        >
          Â£
          100
           /
           
          Month
        </span>
      </div>
      <div
        class="w-full h-1 bg-[#DDDDDD] rounded-full"
      >
        <div
          class="h-full bg-[#64D98A] rounded-full"
          data-testid="progress-bar"
          style="width: 50%;"
        />
      </div>
    </div>
  </div>
</div>
`;
